system/containers.toit

```diff
@@ -118,18 +118,98 @@ class ContainerImageFlash extends ContainerImage:
     // TODO(kasper): Check in the other methods before using this?
     allocation_ = null
 
-class ContainerManager implements SystemMessageHandler_:
+abstract class ContainersServiceDefinition extends ServiceDefinition
+    implements ContainersService:
+  constructor:
+    super ContainersService.NAME
+        --major=ContainersService.MAJOR
+        --minor=ContainersService.MINOR
+    install
+
+  handle client/int index/int arguments/any -> any:
+    if index == ContainersService.LIST_IMAGES_INDEX:
+      return list_images
+    if index == ContainersService.START_IMAGE_INDEX:
+      return start_image (uuid.Uuid arguments)
+    if index == ContainersService.UNINSTALL_IMAGE_INDEX:
+      return uninstall_image (uuid.Uuid arguments)
+    if index == ContainersService.IMAGE_WRITER_OPEN_INDEX:
+      return image_writer_open client arguments
+    if index == ContainersService.IMAGE_WRITER_WRITE_INDEX:
+      return image_writer_write client arguments[0] arguments[1]
+    if index == ContainersService.IMAGE_WRITER_COMMIT_INDEX:
+      return (image_writer_commit client arguments).to_byte_array
+    if index == ContainersService.IMAGE_WRITER_CLOSE_INDEX:
+      return image_writer_close client arguments
+    unreachable
+
+  abstract image_registry -> FlashRegistry
+  abstract images -> List
+  abstract add_flash_image allocation/FlashAllocation -> ContainerImage
+  abstract lookup_image id/uuid.Uuid -> ContainerImage?
+
+  abstract register_descriptor client/int descriptor/Object -> int
+  abstract lookup_descriptor client/int handle/int -> any
+  abstract unregister_descriptor client/int handle/int -> none
+
+  list_images -> List:
+    return images.map --in_place: | image/ContainerImage |
+      image.id.to_byte_array
+
+  start_image id/uuid.Uuid -> int?:
+    image/ContainerImage? := lookup_image id
+    if not image: return null
+    return image.start.id
+
+  uninstall_image id/uuid.Uuid -> none:
+    image/ContainerImage? := lookup_image id
+    if not image: return
+    image.delete
+
+  image_writer_open size/int -> int:
+    unreachable  // <-- TODO(kasper): Nasty.
```

Comment:
Why are they unreachable?

Reply:
They are so that the service definition appears to implement the **client** interface. I'm working on a way to improve this.